### PR TITLE
feat: show source code lines / traceback from `ReceiptAPI` [APE-708]

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

fixes: #867 
Fixes: APE-217
Fixes: APE-577

* Adds `traceback` property to `ReceiptAPI` with implementation in `ape_ethereum`
* Adds `show_source_traceback()` method to `ReceiptAPI` the same.

### How I did it

* Using AST, trace execution, and PCMap. Only works with Vyper at first.

### How to verify it

Create `interfaces/IRegistryF.vy`:

```python
# @version 0.3.7

@external
def register_f(addr: address):
    pass
```

Create `RegistryF.vy`:

```python
# @version 0.3.7

addr: public(address)

@external
def register_f(addr2: address):
    assert self.addr != addr2, "doubling."
    self.addr = addr2
```

Create `JustFail.vy`:

```python
# @version 0.3.7

import interfaces.IRegistryF as IRegistryF

_balance: public(uint256)
registry: public(IRegistryF)

@external
def __init__(registry: IRegistryF):
    self.registry = registry

@external
def addBalance_f(
    num: uint256
) -> uint256:
    assert num != self._balance
    self.registry.register(msg.sender)
    self._balance = self._balance + self.addInterest(num)

    # Run some loops.
    for i in [1, 2, 3, 4, 5]:
        if i == num:
            break

    # Fail in the middle (is test)
    # Fails because was already set above.
    self.registry.register_f(msg.sender)

    for i in [1, 2, 3, 4, 5]:
        if i != num:
            continue

    return self._balance

@internal
def addInterest(num: uint256) -> uint256:
    return 123 + num
```

Create a script `tb.py`:

```python
import ape
from ape.exceptions import ContractLogicError


def main():
    account = ape.accounts.test_accounts[0]
    registry = ape.project.RegistryF.deploy(sender=account)
    contract = ape.project.JustFail.deploy(registry, sender=account)
    txn = contract.addBalance_f.as_transaction(123)

    try:
        account.call(txn)
    except ContractLogicError:
        pass

    receipt = txn.provider.get_receipt(txn.txn_hash.hex())
    receipt.show_source_traceback()
```

It should show:

```shell
Traceback (most recent call last)
  File /home/ApeProjects/ape-playground/contracts/JustFail.vy, in addBalance_f(uint256 num) -> uint256
       20     # Run some loops.
       21     for i in [1, 2, 3, 4, 5]:
       22         if i == num:
       23             break
       24
       25     # Fail in the middle (is test)
       26     # Fails because was already set above.
  -->  27     self.registry.register_f(msg.sender)
       28
       29     for i in [1, 2, 3, 4, 5]:
       30         if i != num:
       31             continue

  File /home/ApeProjects/ape-playground/contracts/RegistryF.vy, in register_f(address addr2)
  -->  7     assert self.addr != addr2, "doubling."
       8     self.addr = addr2
```

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape_ethereum/transactions.py [hunk 1]
```diff
@@ -157,17 +188,166 @@ def call_tree(self) -> Optional[CallTreeNode]:
...
         method_id = self.data[:4]
-        if method_id not in contract_type.methods:
+        if method_id not in self.contract_type.methods:
             return None
 
-        return contract_type.methods[method_id]
+        return self.contract_type.methods[method_id]
+
+    @property
+    def traceback(self) -> Traceback:
```
Comment (unparalleled-js) [inline src/ape_ethereum/transactions.py:210]: The chaos happens here

File: src/ape/api/transactions.py [hunk 2]
```diff
@@ -402,6 +413,14 @@ def show_gas_report(self, file: IO[str] = sys.stdout):
         Display a gas report for the calls made in this transaction.
         """
 
+    @raises_not_implemented
+    def show_traceback(self):
```
Comment (fubuloubu) [inline src/ape/api/transactions.py:417]: I think `show_trace` is a little to close to this name, maybe `show_error_traceback` (and corresponding `error_traceback`)

File: src/ape/types/trace.py [hunk 3]
```diff
@@ -219,3 +221,150 @@ class TraceFrame(BaseInterfaceModel):
     """
     The raw trace frame from the provider.
     """
+
+
+class Traceback(list):
```
Comment (fubuloubu) [inline src/ape/types/trace.py:226]: With pydantic, you can do this:

```py
class Traceback(BaseModel):
    __root__: List[TracebackItem]

    ...
```

Comment (unparalleled-js) [reply src/ape/types/trace.py:226]: That isn't quite the same! They are still models wrapped around it and not a subclass directly. Example:

```python
from pydantic import BaseModel
from typing import List

class MyList(BaseModel):
    __root__: List[int] = []

class MyStr(BaseModel):
    __root__: str = ""

my_ls = MyList()
my_str = MyStr()
print(f"has append: {hasattr(my_ls, 'append')}")
print(f"str of custom str: {my_str}")
```

Outputs:

```
has append: False
str of custom str: __root__=''
```

This is why in `ethpm-types`, we had to add the `__str__` and `__repr__` methods still when we did `__root__: str`, 'member?

File: src/ape/api/transactions.py (hunk 2 above)
Comment (unparalleled-js) [reply src/ape/api/transactions.py:417]: I agree. However, I was thinking - there is no reason to not have this option for non-failing receipts.. Like it still works, you just make it all the way to the `RETURN` opcode in the traceback. We can make it not be red or something.

**Comment by unparalleled-js** [comment]
- [ ] Main thing this needs is thorough testing, both negative and positive.

**Comment by unparalleled-js** [comment]
Observation: would like to get it like IPython where it selectively shows a few lines before or after:'

```
File ~/PycharmProjects/ape/src/ape/managers/chain.py:540, in AccountHistory.__getitem_slice(self, indices)
    537     stop += len(self)
    539 elif stop > len(self):
--> 540     raise ChainError(
    541         f"'stop={stop}' cannot be greater than account's current nonce ({len(self)})."
    542     )
    544 if stop <= start:
    545     return []  # nothing to query
```

I wonder how it does that...

**Comment by fubuloubu** [comment]
> Observation: would like to get it like IPython where it selectively shows a few lines before or after:'
> 
> ```
> File ~/PycharmProjects/ape/src/ape/managers/chain.py:540, in AccountHistory.__getitem_slice(self, indices)
>     537     stop += len(self)
>     539 elif stop > len(self):
> --> 540     raise ChainError(
>     541         f"'stop={stop}' cannot be greater than account's current nonce ({len(self)})."
>     542     )
>     544 if stop <= start:
>     545     return []  # nothing to query
> ```
> 
> I wonder how it does that...
> 
> * [ ]  Might need to delegate some of this to the `CompilerAPI` but need to figure out what exactly...

Could maybe find the next 1-2 layer higher AST nodes and display the breadth of those

File: src/ape/api/transactions.py [hunk 4]
```diff
@@ -385,6 +385,17 @@ def return_value(self) -> Any:
...
+    @property
+    @raises_not_implemented
+    def traceback(self) -> Traceback:
+        """
+        A pythonic style traceback for both failing and non-failing receipts.
+        Requires a provider that implements
+        :meth:~ape.api.providers.ProviderAPI.get_transaction_trace`.
+        """
+
+        return  # type: ignore[return-value]
```
Comment (fubuloubu) [inline src/ape/api/transactions.py:397]: Do you need a `return` statement w/ `raises_not_implemented`?

Comment (unparalleled-js) [reply src/ape/api/transactions.py:397]: If there is a return it needs something, but the other cases we do the type ignore in the signature without the return, which is better, so I will switch it to that

Comment (unparalleled-js) [reply src/ape/api/transactions.py]: done, but has to squash commits because of hefting rebasing (easier to rebase 1 commit), but the change is in there.

File: src/ape/types/trace.py (hunk 3 above)
Comment (unparalleled-js) [reply src/ape/types/trace.py]: I ended up doing this!

**Comment by unparalleled-js** [comment]
- [ ] still trying to find way to exclude irrelevant stack frames when err happens in console

File: src/ape/_cli.py [hunk 5]
```diff
@@ -36,14 +46,17 @@ def invoke(self, ctx) -> Any:
...
-            if logger.level == LogLevel.DEBUG.value:
-                tb = traceback.format_exc()
-                err_message = tb or str(err)
-            else:
-                err_message = str(err)
+        except TransactionError as err:
+            if not err.traceback:
+                raise abort(err) from err
+
+            reporter = ctx.obj.chain_manager._reports
```
Comment (unparalleled-js) [inline src/ape/_cli.py:53]: - [ ] Either use the `__traceback__` that gets set now, or manually include the frame(s) from the script or plugin making the call that causes the tx err

**Comment by unparalleled-js** [comment]
- [ ] Show in pytest plugin

**Comment by unparalleled-js** [comment]
- [ ] Figure out why isn't working on `ape-foundry` and `ape-hardhat` (only been using `ape-geth` thus far)

**Comment by unparalleled-js** [comment]
Marking this as ready for review... the PR is gotten unwieldly, but this feels ready now.

* Can see awesome stacktraces in scripts
* Can see awesome stacktraces in IPython
* Can see awesome stacktraces in Pytest
* Can see (and interact with) awesome stacktraces when using debuggers.

Future needs:

* Compiler-specific handling pieces (will look into that next)
* Scripting, pytest, and console integration tests

**Comment by unparalleled-js** [comment]
- [ ] Compiler API delegation
- [ ] Handle implicit payable checks in Vyper

File: src/ape/types/trace.py [hunk 6]
```diff
@@ -19,6 +22,9 @@
 """
 
 
+_FUNCTION_AST_TYPES = ("FunctionDef", "Name", "arguments")
```
Comment (unparalleled-js) [inline src/ape/types/trace.py:25]: - [ ] find way to delegate to `CompilerAPI`

File: src/ape/types/trace.py [hunk 7]
```diff
@@ -325,7 +325,7 @@ def get_content_asts(self, location: SourceLocation) -> List[ASTNode]:
         return [
             a
             for a in self.ast.get_nodes_at_line(location)
-            if a.lineno >= location[0] and a.classification == ASTClassification.CONTENT
+            if a.lineno >= location[0] and a.classification != ASTClassification.FUNCTION_DEF
```
Comment (fubuloubu) [inline src/ape/types/trace.py:328]: I think this also works, and maybe be suggested for enums
```suggestion
            if a.lineno >= location[0] and a.classification is ASTClassification.FUNCTION_DEF
```

File: src/ape/managers/chain.py [hunk 8]
```diff
@@ -622,15 +624,24 @@ def __getitem_str(self, account_or_hash: str) -> Union[AccountHistory, ReceiptAP
...
+
+            # If we get here, we failed to get an account or receipt.
+            # Raise top-level exception.
+            raise
 
-            return receipt
+    def _get_receipt(self, txn_hash: str) -> ReceiptAPI:
+        receipt = self._hash_to_receipt_map.get(txn_hash)
+        if not receipt:
+            # TODO: Replace with query manager once supports receipts
```
Comment (fubuloubu) [inline src/ape/managers/chain.py:639]: Technically it supports receipts right now, it's just misnamed as transactions

Comment (unparalleled-js) [reply src/ape/managers/chain.py:639]: what is the best way to query a receipt by its known hash?

File: src/ape/exceptions.py [hunk 9]
```diff
@@ -113,9 +114,82 @@ def __init__(
...
+
+        if not stack:
+            return
+
+        tb_next = None
+        for tb in stack:
+            tb.tb_next = tb_next
+            tb_next = tb
+
+        self.__traceback__ = stack[-1]
```
Comment (fubuloubu) [inline src/ape/exceptions.py:191]: Sorta wish this code was a utility function

Comment (antazoey) [reply src/ape/exceptions.py:191]: i can move it to a method in this same module but cant put it in utils because of circular deps

Comment (fubuloubu) [reply src/ape/exceptions.py]: That's good enough

File: src/ape/types/trace.py [hunk 10]
```diff
@@ -225,3 +229,768 @@ class TraceFrame(BaseInterfaceModel):
...
+        return self.__root__.items()
+
+    def as_list(self) -> List[str]:
+        return list(self.__root__.values())
+
+    def __getitem__(self, lineno: int) -> str:
+        return self.__root__[lineno]
+
+    def __iter__(self):
+        yield from self.__root__
```
Comment (fubuloubu) [inline src/ape/types/trace.py:267]: this would iter the line numbers?

**Review by fubuloubu — DISMISSED** [review]
some of the items under `src/ape/types/trace.py` appear like they could live separately under `ethpm-types`, but not sure I want to advocate for that too much

File: src/ape/types/trace.py (hunk 10 above)
Comment (antazoey) [reply src/ape/types/trace.py]: updated this to be cooler and still work

**Comment by antazoey** [comment]
> some of the items under `src/ape/types/trace.py` appear like they could live separately under `ethpm-types`, but not sure I want to advocate for that too much

OK I did manage to share some code there and move some of the basic structures there. they're composed of ethpm types anyway so it makes sense and `Content` can be used on `Source` to share some indexing logic there, so long that the basic spec can still be honored in serialization. but it is kind of a breaking change in `ethpm-types`, so we will have to do an ethpm-types 0.5 to proceed. it does help separate of the concerns and makes this PR a little smaller too

File: src/ape/types/trace.py [hunk 11]
```diff
@@ -225,3 +230,428 @@ class TraceFrame(BaseInterfaceModel):
     """
     The raw trace frame from the provider.
     """
+
+
+class ControlFlow(BaseModel):
```
Comment (fubuloubu) [inline src/ape/types/trace.py:235]: Does this class have ape-specific functionality?

**Review by fubuloubu — COMMENTED** [review]
looking much cleaner now

File: src/ape/types/trace.py (hunk 11 above)
Comment (antazoey) [reply src/ape/types/trace.py:235]: i thought about moving to `ethpm-types` but decided against it because it really is only about execution. it felt weird putting anything that is solely about execution in ethpm-types for some reason, probably because there wasnt good spot for it on its own.

but the answer is no, it is basically a class that only works with classes from ethpm-types

Comment (antazoey) [reply src/ape/types/trace.py:235]: it is kindof like the main helper class to the source traceback too, they go kind of hand in hand and make the most sense together, at least that is what i think currently
